### ISSUE TITLE
Crashes on Render

### ISSUE THREAD

**Issue by droptopgokart** [xLightsSequencer/xLights#4062]
Crashes on 2023.17 when Rendering

[xLightsProblem.zip](https://github.com/smeighan/xLights/files/12968723/xLightsProblem.zip)
[xLights_dbgrpt-13596-20231017T180412.zip](https://github.com/smeighan/xLights/files/12968724/xLights_dbgrpt-13596-20231017T180412.zip)

**Comment by dartunder** [comment]
I duplicated this in zoom room... 
I'm looking briefly and so far removing the right half of effects ... still crashes
removing the maxtix . it rendered complete.

**Comment by droptopgokart** [comment]
Progress! So is there something that I did that broke it?

**Comment by dartunder** [comment]
I'm not sure how to help more.  @keithsw1111 I tried but maybe you can help?

**Comment by dartunder** [comment]
Thanks Scott, but I compiled your commit, and that sequence still crashes when running some windows only code.

Exception thrown at 0x00007FF7CEFAE781 in Xlights.exe: 0xC0000005: Access violation reading location 
  gc = wxGraphicsRenderer::GetDirect2DRenderer()->CreateContext(*dc);
  
 additional context from RenderBuffer.cpp

#ifdef __WXMSW__
    // must use the Direct2D renderer to get the color emoji's
    // along with custom build of wxWidgets until
    // https://trac.wxwidgets.org/ticket/19275#ticket
    // is applied/fixed
    static std::mutex singleThreadContextCreate;
    {
        // we need to lock before we do this otherwise we try to load libraries concurrently on many threads which causes problems
        std::unique_lock<std::mutex> locker(singleThreadContextCreate);
        gc = wxGraphicsRenderer::GetDirect2DRenderer()->CreateContext(*dc);
        if (gc != nullptr) {
            gc->SetPen(wxPen()); // we set a pen just to ensure the renderer is initialised ... if we dont do this there are scenarios where deleting it can run into issues
        }
    }
#else

**Comment by dartunder** [comment]
I noticed that the name of the font has spaces.

**Comment by computergeek1507** [comment]
Its the Shape Effect with emoji at 34 secs but I don't know how to fix it, above my knowledge...

**Comment by dartunder** [comment]
yeah the name is not surrounded in quotes like it is in a new sequence

**Comment by dartunder** [comment]
FYI, the code references this link:
https://trac.wxwidgets.org/ticket/19275#ticket
Which has been moved to:
https://github.com/wxWidgets/wxWidgets/issues/19275

**Comment by dartunder** [comment]
@dkulp it appears you were involved with the initial fix referenced.
https://github.com/wxWidgets/wxWidgets/issues/19275
https://github.com/wxWidgets/wxWidgets/issues/22167

**Comment by droptopgokart** [comment]
So I'm trying to go in and replace the emoji and morph effects with something else that won't crash, but I cannot keep xlights alive long enough to make the changes.  Any ideas?

**Comment by dartunder** [comment]
before you change anything click the disable render checkbox at the bottom  change it save to new name then try render

**Comment by cybercop23** [comment]
Is this still an issue or can it be closed?

**Comment by derwin12** [comment]
Confirmed .. still an issue .. waiting on upstream fixes in wxwidgets. Perhaps we can fix it in our code?

**Comment by derwin12** [comment]
If I remove this "workaround" and just run the else portion .. the provided sequence runs fine. 
Perhaps updates in wxWidgets have got this corrected? @dkulp 
```
#ifdef __WXMSW__
    // must use the Direct2D renderer to get the color emoji's
    // along with custom build of wxWidgets until
    // https://trac.wxwidgets.org/ticket/19275#ticket
    // is applied/fixed
    static std::mutex singleThreadContextCreate;
    {
        // we need to lock before we do this otherwise we try to load libraries concurrently on many threads which causes problems
        std::unique_lock<std::mutex> locker(singleThreadContextCreate);
        gc = wxGraphicsRenderer::GetDirect2DRenderer()->CreateContext(*dc);
        if (gc != nullptr) {
            gc->SetPen(wxPen()); // we set a pen just to ensure the renderer is initialised ... if we dont do this there are scenarios where deleting it can run into issues
        }
    }
#else
    gc = wxGraphicsContext::Create(*dc);
#endif
```

**Comment by dkulp** [comment]
It's likely fixed due to the addition of the call to wxGraphicsRenderer::GetDirect2DRenderer(); in xLightsMain initializing the direct2d renderer.     That said, make sure the call to wxGraphicsContext::Create(*dc) is actually creating the Direct2D renderer and not the GDI renderer.   The color emoji's definitely need the Direct2D.

**Comment by derwin12** [comment]
Right .. so it doesnt crash .. but it then doesnt show the colored emojis. I guess we leave it as is and wait on wxWidgets.

**Comment by dkulp** [comment]
You could also try changing the shape effect to render on the main thread instead of the background thread.   That may work.

**Comment by derwin12** [comment]
One effect, minimal example.

[crash shape effect.zip](https://github.com/user-attachments/files/19688484/crash.shape.effect.zip)

**Comment by derwin12** [comment]
What was originally thought to be a wxwidget issue, the original sequence had an emoji effect on a whole house set to Single Line. The resulting buffer was too big to create the context and would crash. I have set the buffer limit to something reasonable so it doesn't crash. Changing the group's Single Line render style to Max Grid or using Per Preview would have avoided this crashing condition as well.
